editor/src/components/navigator/navigator-item/component-picker.tsx

```diff
@@ -59,20 +59,90 @@ export function componentPickerOptionTestId(componentName: string, variant?: str
 }
 
 export const ComponentPicker = React.memo((props: ComponentPickerProps) => {
+  const { onItemClick } = props
+  const [selectedComponentKey, setSelectedComponentKey] = React.useState<string | null>(null)
   const [filter, setFilter] = React.useState<string>('')
+  const menuRef = React.useRef<HTMLDivElement | null>(null)
 
-  const allComponentsToShow: InsertMenuItemGroup[] = []
-
-  props.allComponents.forEach((c) => {
-    allComponentsToShow.push({
-      ...c,
-      options: c.options.filter((o) =>
-        o.label.toLocaleLowerCase().includes(filter.toLocaleLowerCase().trim()),
-      ),
+  const componentsToShow = useMemo(() => {
+    const allComponentsToShow: InsertMenuItemGroup[] = []
+    props.allComponents.forEach((c) => {
+      allComponentsToShow.push({
+        ...c,
+        options: c.options.filter((o) =>
+          o.label.toLocaleLowerCase().includes(filter.toLocaleLowerCase().trim()),
+        ),
+      })
     })
-  })
+    return allComponentsToShow
+  }, [props.allComponents, filter])
+
+  const flatComponentsToShow = useMemo(() => {
+    return componentsToShow.flatMap((c) => c.options)
+  }, [componentsToShow])
+
+  const highlightedComponentKey = useMemo(() => {
+    const firstOptionKey =
+      flatComponentsToShow.length > 0 ? flatComponentsToShow[0].value.key : null
+    if (selectedComponentKey == null) {
+      return firstOptionKey
+    }
+    // check if selectedComponentKey is still in the list
+    const found = flatComponentsToShow.some((c) => c.value.key === selectedComponentKey)
+    return found ? selectedComponentKey : firstOptionKey
+  }, [flatComponentsToShow, selectedComponentKey])
+
+  const onItemHover = useCallback(
+    (elementToInsert: InsertMenuItemValue) => {
+      return () => {
+        setSelectedComponentKey(elementToInsert.key)
+      }
+    },
+    [setSelectedComponentKey],
+  )
 
-  const componentsToShow = allComponentsToShow
+  const onKeyDown = useCallback(
+    (e: React.KeyboardEvent<HTMLInputElement>) => {
+      if (e.key === 'ArrowDown') {
+        const currentIndex = flatComponentsToShow.findIndex(
+          (c) => c.value.key === highlightedComponentKey,
+        )
+        if (currentIndex >= 0 && currentIndex < flatComponentsToShow.length - 1) {
+          const newKey = flatComponentsToShow[currentIndex + 1].value.key
```

Comment:
this could be easily factored out to a helper function and just pass the cursor index delta, right?